Mathlib/Data/Nat/EvenOddRec.lean

```diff
@@ -57,4 +56,37 @@ theorem evenOddRec_odd (n : ℕ) (P : ℕ → Sort _) (h0 : P 0) (h_even : ∀ i
   eq_of_heq (this _ (bit1_val _))
 #align nat.even_odd_rec_odd Nat.evenOddRec_odd
 
+/-- Strong recursion principle on even and odd numbers: if for all `i : ℕ` we can prove `P (2 * i)`
+from `P j` for all `j < 2 * i` and we can prove `P (2 * i + 1)` from `P j` for all `j < 2 * i + 1`,
+then we have `P n` for all `n : ℕ`. -/
+noncomputable
+def evenOddStrongRec {p : Sort _} (n : ℕ)
```

Comment:
Surely this should be:
```suggestion
def evenOddStrongRec {p : ℕ → Prop} (n : ℕ)
```
after all:
```
example {p : Sort _}
    (peven : ∀ m : ℕ, (∀ k : ℕ, k < 2 * m → p) → p) : p :=
  peven 0 (fun k hk ↦ by simp at hk)
```

Reply:
Yeah, this was basically my original definition (but with `Sort` rather than `Prop`):
```
noncomputable def evenOddStrongRec' {p : ℕ → Sort _} (n : ℕ)
    (peven : ∀ m : ℕ, (∀ k : ℕ, k < 2 * m → p k) → p (2 * m))
    (podd : ∀ m : ℕ, (∀ k : ℕ, k < 2 * m + 1 → p k) → p (2 * m + 1)) : p n :=
n.strong_rec_on <| fun m ih => m.even_or_odd'.choose_spec.by_cases
  (fun hm => by rw [hm] at ih ⊢; exact peven _ ih)
  (fun hm => by rw [hm] at ih ⊢; exact podd _ ih)
```
But with this I can't seem to make recursive definitions, which I can with the current definition:
```
noncomputable def foo : ℕ → ℕ :=
  fun n => evenOddStrongRec n
    (fun m => match m with
      | 0 => fun _ => sorry
      | n+1 => fun _ => sorry)
    (fun m => match m with
      | 0 => fun _ => sorry
      | n+1 => fun _ => sorry)
```
Interestingly, `Nat.strong_rec_on` isn't cased as `Nat.strongRecOn`?

Reply:
I apologise if I'm being dense but I am still unable to understand why `evenOddStrongRec` is useful. Doesn't the `example` I provided above show that it's not really saying anything useful?

Reply:
Yes, but my point is this was a hack so I can define my `foo`. I changed the definition but now I can't seem to prove `_even` nor `_odd`. Help wanted!

Reply:
> I'm not going to bother with `_even` and `_odd` because I can't prove them, but the definition is enough for me.

Lean aside, it was never clear to me what the alternative meant. What you have now looks good though :-)